### ISSUE TITLE
Panic calling current_thread::block_on_all

### ISSUE THREAD

**Issue by rphmeier** [tokio-rs/tokio#410]
I've got this test harness code that leads to a panic. I don't know if there's a better way to test stuff using the tokio runtime but this was my best attempt so far. 

Originally I thought the errors were because of panics inside the futures running on tokio, so a lot of this harness is just to get us the result after execution so we can do further testing on the value/error.

With tokio 0.1.7.

```rust
fn test_harness<F, X, T, U>(f: F) -> Result<T, U> where
	F: FnOnce() -> X + Send + 'static,
	X: IntoFuture<Item=T,Error=U>,
	X::Future: 'static,
	T: Send + 'static,
	U: Send + 'static,
{

	let maybe_err = Arc::new(Mutex::new(None));
	let runtime_err_handle = maybe_err.clone();

	runtime::run(future::lazy(move || {
		let to_spawn = f().into_future();
		let res = ::tokio::executor::current_thread::block_on_all(to_spawn);
		*runtime_err_handle.lock().unwrap() = Some(res);
		Ok(())
	}));

	let x: Option<Result<T, U>> = maybe_err.lock().unwrap().take();

	x.expect("runtime concluded, value set")
}
```

Relevant excerpt of backtrace:

```
thread 'tokio-runtime-worker-0' panicked at 'called `Result::unwrap()` on an `Err` value: EnterError { _a: () }'
  10: <tokio::executor::current_thread::CurrentThread<P>>::block_on
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.7/src/executor/current_thread/mod.rs:362
  11: substrate_bft::generic::tests::test_harness::{{closure}}
             at substrate/bft/src/generic/tests.rs:228
  12: <futures::future::lazy::Lazy<F, R>>::get
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/future/lazy.rs:64
  13: <futures::future::lazy::Lazy<F, R> as futures::future::Future>::poll
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/future/lazy.rs:82
  14: <alloc::boxed::Box<F> as futures::future::Future>::poll
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/future/mod.rs:113
  15: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/task_impl/mod.rs:289
  16: <futures::task_impl::Spawn<T>>::enter::{{closure}}
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/task_impl/mod.rs:363
  17: futures::task_impl::std::set
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/task_impl/std/mod.rs:78
  18: <futures::task_impl::Spawn<T>>::enter
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/task_impl/mod.rs:363
  19: <futures::task_impl::Spawn<T>>::poll_future_notify
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.21/src/task_impl/mod.rs:289
  20: tokio_threadpool::task::TaskFuture::poll
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/task/mod.rs:292
  21: tokio_threadpool::task::Task::run::{{closure}}
             at /Users/robert/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/task/mod.rs:165
```

Which corresponds to this code:

```rust
impl current_thread::Executor {
    // ...
    pub fn block_on<F>(&mut self, future: F)
        -> Result<F::Item, BlockError<F::Error>>
    where F: Future
    {
        let mut enter = tokio_executor::enter().unwrap(); // L362: panic is this unwrap
        self.enter(&mut enter).block_on(future)
    }
    // ...
}
```

**Comment by rphmeier** [comment]
(off-topic a bit)

For a codebase as large as complex as tokio I think it's crucial to enforce a stricter style around panickers and `unwrap` in particular. 

The rule of thumb I follow is "prove or remove" where I use `expect` and `panic` with explicit strings that describe exactly the preconditions that need to be met and proves that they are not met in order to show that the panic is unreachable.

Consider

```rust
my_optional_val.expect("Always set if property X holds; why X holds; qed");
```

vs.

```rust
my_optional_val.unwrap();
```

I think this rule gives a lot of benefits for readability and debugging, since you can always see what's supposed to make a panic not happen, and in the odd chance something does break, the panic backtrace will tell you exactly which assumptions failed.

I can extract this discussion out into another issue if desired

**Comment by rphmeier** [comment]
If this _is_ the intended behavior, then I really hope that it gets changed at some point. In my application I'd like to run some kinds of work on any thread in the thread pool and have some tasks spawn new non-Send work onto the current thread. The `executor::current_thread` API seems like the only way to do that so far.

**Comment by kpp** [comment]
> in practice I would want to have `tokio::spawn`

Your code semantically is blocking that's why you have to use `block_on`.

You may use tokio::spawn, but instead of maybe_err(Mutex) you would like to use [oneshot channel](https://docs.rs/futures/*/futures/channel/oneshot/index.html), send the result of execution into tx part and wait on rx part right where you exactly need it. Or even like that: https://github.com/tokio-rs/tokio/blob/master/tokio-threadpool/examples/hello.rs

**Comment by rphmeier** [comment]
I agree the mutex is probably not the right way to go although for a test harness it's fine to be a little rough around the edges.

Note that blocking at the I/O level is independent from the concurrency of tasks provided by a tokio executor, which lets us perform work as soon as some asynchronous work in the background finishes. While synchronization primitives can deadlock in practice, you can also stall tasks with constructs like infinite loops. In this case the error isn't caused by deadlock.

**Comment by kpp** [comment]
OK. Your code in the description message failed:

```
	runtime::run(future::lazy(move || {
		let to_spawn = f().into_future();
		let res = ::tokio::executor::current_thread::block_on_all(to_spawn);
		*runtime_err_handle.lock().unwrap() = Some(res);
		Ok(())
	}));
```

Because `runtime::run` runs a multithreaded executor while you are trying to get the singlethreaded executor (which is None).

**Comment by carllerche** [comment]
I'm not exactly sure what you are trying to do, however, you should never call a `block_on` function from within a future (which you are doing in the lazy block). This is why it panics.

Re: panic messages, I would be happy to take a PR that improves the messaging. I tend to not like `expect` because it also loses the context of the error, but the error value here was not super useful either.

**Comment by kpp** [comment]
how about?

```diff
--- a/tokio-executor/src/enter.rs
+++ b/tokio-executor/src/enter.rs
@@ -20,11 +20,16 @@ pub struct Enter {
 
 /// An error returned by `enter` if an execution scope has already been
 /// entered.
-#[derive(Debug)]
 pub struct EnterError {
     _a: (),
 }
 
+impl fmt::Debug for EnterError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "The execution scope has already been entered")
+    }
+}
+
```

**Comment by rphmeier** [comment]
> I tend to not like expect because it also loses the context of the error

this makes sense, at least in contexts where errors are expected to happen. In that case, unwrap is actually worse. IMO It's important to make a distinction between locations where errors _can_ happen and where errors are logically impossible to happen. 

Most of the unwraps I've seen in the tokio codebase are meant to denote the latter, and in those cases I think the "prove or remove" rule makes sense.

For cases like this where the panic occurs as a result of user error I think it's most important to document under exactly what situations the panic can occur so users can avoid it.

**Comment by carllerche** [comment]
@kpp 

I would probably change the impl to something like:

```rust
f.debug_struct("EnterError")
    .field("reason", &"attempted to run an executor while another executor is already running")
    .finish()
```

(or something).

And the `unwrap` would probably do something like:

```rust
if let Err(e) = enter() {
     panic!("failed to start `current_thread::Runtime`; {:?}", e);
}
```
